Coverage tests for remaining TerminatorKinds and async, improve Assert

richkadel opened the pull request:
Tested and validate results for panic unwind, panic abort, assert!()
macro, TerminatorKind::Assert (for example, numeric overflow), and
async/await.

Implemented a previous documented idea to change Assert handling to be
the same as FalseUnwind and Goto, so it doesn't get its own
BasicCoverageBlock anymore. This changed a couple of coverage regions,
but I validated those changes are not any worse than the prior results,
and probably help assure some consistency (even if some people might
disagree with how the code region is consistently computed).

Fixed issue with async/await. AggregateKind::Generator needs to be
handled like AggregateKind::Closure; coverage span for the outer async
function should not "cover" the async body, which is actually executed
in a separate "closure" MIR.

rust-highfive commented:
r? @oli-obk

(rust_highfive has picked a reviewer for you, use r? to override)

richkadel commented:
r? @tmandry

richkadel commented:
fyi: @wesleywiser

richkadel commented:
! ... I forgot to bless at least one of the new tests. I'm building the blessed results now and will push an update.

Then I also plan to pull this PR down to my Mac and to my Windows VM so I can make sure (as best as I can) that these tests work on all 3 platforms (including Linux, my main dev environment).

richkadel commented:
Oh, and I just realized I should rebase with the latest changes (including the 2 PRs that just landed for coverage). They may not affect this PR, but I need to be sure.

richkadel commented:
new tests are blessed and pass on Linux... I'll be testing Mac and Windows next.

richkadel commented:
FYI, something seems broken in the rustc build. After rebasing today, and then pulling down this PR to MacOS and Windows local environments, neither of those will build the `profiler_builtins` library anymore.

https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/profiler_builtins.20not.20compiling.20on.20MacOS.20or.20Windows

I "think" it might still work on Linux, but I'm a little afraid to try a clean build at this point 😨

richkadel commented:
Issue #79124 affects local builds of this PR

richkadel commented:
Tests are working on Mac. Windows test is taking forever... ;-)

richkadel commented:
Windows tests pass as well.

@tmandry - subject to any review feedback changes requested, this PR should be good to merge.

Thanks!

tmandry commented on src/test/run-make-fulldeps/coverage-reports-base/expected_show_coverage.async.txt:
```diff
@@ -0,0 +1,66 @@
+   25|      1|async fn i(x: u8) {
+   26|      1|    match x {
+   27|      1|        y if f().await == y + 1 => (),
+                      ^0                         ^0
```
Does the `y` being uncovered also happen in non-async code? I think I remember you mentioning something like this. It's not wrong if you consider that it's the binding of `y`. It is a bit unintuitive, though.

richkadel replied:
Yes, this is the same for async and non-async.

Yes it was discussed because it is less intuitive. At one point, I tried to include the `y` in the pattern code region, but the logic to allow that (which may seem more intuitive, but is counter to what the MIR says) caused problems handling other cases.

I decided it was better to honor the MIR in all cases and show what's really happening.

This is not *just* one of those situations where we can be accurate but non-intuitive, or intuitive but inaccurate; the other problem is, it would require special case handling in the coverage code (which is something we both prefer to avoid). 😄

tmandry replied:
Yeah, special casing is not great. We can always change the way MIR building emits source info at some point.

tmandry commented on src/test/run-make-fulldeps/coverage-reports-base/expected_show_coverage.async.txt:
```diff
@@ -0,0 +1,66 @@
+   59|      1|            if let Poll::Ready(val) = future.as_mut().poll(&mut context) {
+   60|      1|                break val;
+   61|       |            }
+   62|      0|        }
```
This would cause tooling to complain about missing coverage of this line, but the break does happen.

Would it be possible to make all loop exits count the closing loop brace?

richkadel replied:
@tmandry - I'm confused by your interpretation on this one.

To me, this is intuitive. The bottom of the loop is never reached.

This is a coverage gap because the program does not test an iteration of the loop when it does *not* break.

tmandry replied:
Oh, now I see. That makes sense.

tmandry commented on src/test/run-make-fulldeps/coverage-reports-base/expected_show_coverage.conditions.txt:
```diff
@@ -36,34 +37,70 @@
-   54|      1|        return;
-   55|       |    };
+   39|      1|    if true {
+   40|       |        // Demonstrate the difference with `TerminatorKind::Assert` as of 2020-11-15. Assert is no
+   41|       |        // longer treated as a `BasicCoverageBlock` terminator, which changed the coverage region,
+   42|       |        // for the executed `then` block above, to include the closing brace on line 30. That
+   43|       |        // changed the line count, but the coverage code region (for the `else if` condition) is
+   44|       |        // still valid.
+   45|       |        //
+   46|       |        // Note that `if` (then) and `else` blocks include the closing brace in their coverage
+   47|       |        // code regions when the last line in the block ends in a semicolon, because the Rust
+   48|       |        // compiler inserts a `StatementKind::Assign` to assign `const ()` to a `Place`, for the
+   49|       |        // empty value for the executed block. When the last line does not end in a semicolon
+   50|       |        // (that is, when the block actually results in a value), the additional `Assign` is not
+   51|       |        // generated, and the brace is not included.
+   52|      1|        let mut countdown = 0;
```
Yeah, this is pretty unintuitive. "Closing brace is counted when the scope is exited, or skipped by if/loop condition," seems like the most logical behavior to me.

richkadel replied:
> "Closing brace is counted when the scope is exited, or skipped by if/loop condition,"

@tmandry are you quoting something here, or just suggesting a way you'd like it described?

On the one hand, counting the closing brace when exiting the scope is how we currently count the closing brace of a function, so this would be more consistent with that.

An argument could be made that I shouldn't count the closing brace of a function, though, if the function was exited by an early `return`.

The truth is, in some cases, the reason things work the way they do for functions and for loops is because it was the best I could do given the MIR, and seemed "reasonable" (even if there were alternatives like these).

Personally, I like the idea of changing how function end braces are counted, rather than changing how loop end braces are counted. I think it's more intuitive to count the end brace only if the code did not exit the scope earlier (by break, continue, or return).

Either way, it's inconsistent now.

Should we continue debating this here? Or log it as an unresolved new "Issue" and invite the community for comment?

If the latter, maybe I shouldn't change either one, for now, since it could be a waste of time if we decide to go a different way.

tmandry replied:
Yeah I think there are a few competing principles..

- Consistency
- Intuitiveness
- Preserving all information

and I'm not exactly sure how to resolve them. My quote was one idea I had.

This might be good to hash out in an issue somewhere.

richkadel replied:
I think a lot of things like this are resolved with my most recent improvements.

tmandry commented on src/test/run-make-fulldeps/coverage-reports-base/expected_show_coverage.panic_unwind.txt:
```diff
@@ -0,0 +1,50 @@
+   42|       |//   5. The reason the coverage results actually show `panic!()` was called is most likely because
+   43|       |//      `panic!()` is a macro, not a simple function call, and there are other `Statement`s and/or
+   44|       |//      `Terminator`s that execute with a coverage counter before the panic and unwind occur.
+   45|       |//   6. By best practice, programs should not panic. By design, the coverage implementation will not
+   46|       |//      incur additional cost (in program size and execution time) to improve coverage results for
+   47|       |//      an event that is not supposted to happen.
```
I don't think this is _always_ true. Of course there are uses like when invariants are violated (bugs), and `#[should_panic]` tests which you mention below. In addition, [salsa](https://github.com/salsa-rs/salsa) supports [unwinding as a cancellation mechanism](https://docs.rs/salsa/0.16.0/salsa/struct.Runtime.html#method.is_current_revision_canceled) for queries.

richkadel replied:
Point taken. It's still my opinion that the *best* practice is not to use `panic!` as a programming model. (Even salsa appears to support both `panic` and `Result::Err` for cancelling, if I'm not mistaken. I'm curious why not just go with the Result::Err method, unless it's for legacy reasons.)

But since this may be more my opinion, I'll rephrase.

tmandry replied:
I think they support unwinding because otherwise cancellation becomes "infectious" and something you have to handle throughout your code, even though it isn't a classic error handling case. I think rust-analyzer uses it.

richkadel replied:
I'm updating the comment.

richkadel replied:
fixed

tmandry commented on src/test/run-make-fulldeps/coverage-reports-base/expected_show_coverage.yield.txt:
```diff
@@ -0,0 +1,38 @@
+   19|      0|        _ => panic!("unexpected value from resume"),
+   20|       |    }
+   21|       |
+   22|      1|    let mut generator = || {
+   23|      1|        yield 1;
+   24|      1|        yield 2;
+   25|      0|        yield 3;
+   26|      0|        return "foo"
+   27|      0|    };
```
This is cool :D

tmandry commented on src/test/run-make-fulldeps/coverage-reports-deadcode/expected_show_coverage.assert.txt:
```diff
@@ -0,0 +1,34 @@
+    1|       |#![allow(unused_assignments)]
```
I'm not seeing how these -deadcode test files are any different, or add value. I guess that's one piece of the cleanup to be done.

richkadel replied:
The `-Clink-dead-code` defaults to `on` on Linux and Mac, and `off` on Windows. If the results are different, the tests would fail on Windows if I only had one variation.

I wasn't sure if I'd get different coverage results depending on the flag. Now that I have several examples, I'll `diff` them to see. It's a little risky to assume the flag makes no difference, even if true for the current tests.

richkadel replied:
@tmandry - FYI, currently, two tests have slightly different coverage results.

```shell
$ diff -r coverage-reports-base coverage-reports-deadcode -x '*counters*' -x '*.json' -x 'prett*' -x 'Make*'
```
```diff
15c15
<    15|       |async fn e() -> u8 { 1 }
---
>    15|      0|async fn e() -> u8 { 1 }
19c19
<    19|       |async fn foo() -> [bool; 10] { [false; 10] }
---
>    19|      0|async fn foo() -> [bool; 10] { [false; 10] }
diff --color -r -x '*counters*' -x '*.json' -x 'prett*' -x 'Make*' '--color=never' coverage-reports-base/expected_show_coverage.partial_eq.txt coverage-reports-deadcode/expected_show_coverage.partial_eq.txt
5a6,15
>   ------------------
>   | Unexecuted instantiation: <partial_eq::Version as core::cmp::PartialOrd>::gt
>   ------------------
>   | Unexecuted instantiation: <partial_eq::Version as core::cmp::PartialOrd>::le
>   ------------------
>   | Unexecuted instantiation: <partial_eq::Version as core::cmp::PartialOrd>::ge
>   ------------------
>   | <partial_eq::Version as core::cmp::PartialOrd>::lt:
>   |    4|      1|#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
>   ------------------
8a19,28
>   ------------------
>   | Unexecuted instantiation: <partial_eq::Version as core::cmp::PartialOrd>::gt::{closure#0}
>   ------------------
>   | Unexecuted instantiation: <partial_eq::Version as core::cmp::PartialOrd>::ge::{closure#0}
>   ------------------
>   | <partial_eq::Version as core::cmp::PartialOrd>::lt::{closure#0}:
>   |    7|      1|    minor: usize,
>   ------------------
>   | Unexecuted instantiation: <partial_eq::Version as core::cmp::PartialOrd>::le::{closure#0}
>   ------------------
```

richkadel replied:
deadcode is no longer applicable to coverage! I'm able to instrument the deadcode functions without using the `-Clink-dead-code` option. Special tests are now gone.

tmandry reviewed (commented):
Switching to focusing on test behavior now.

I think the most important tweak would be to focus on how braces behave, and try to make them as consistent as possible across syntactic forms.

richkadel commented on src/test/run-make-fulldeps/coverage-reports-base/expected_show_coverage.abort.txt:
```diff
@@ -0,0 +1,35 @@
+   16|     11|    while countdown > 0 {
+   17|     10|        if countdown < 5 {
+   18|      4|            might_abort(false);
+   19|      6|        }
```
This is actually correct.

The condition `countdown < 5` executed 10 times (10 loop iterations).

It evaluated to `true` 4 times, and executed the `might_abort()` call.

It skipped the body of the `might_abort()` call 6 times (hidden `else`).

I wanted to capture the coverage of the non-true condition. Say the condition was `countdown < 50`, which is always `true`. In that case, we wouldn't have a test for what happens if `might_abort()` is not called.

So counting the "hidden else" is important.

I can clarify this, but I should probably clarify it in a different test that's more focused on `if` blocks inside loops, and keep this test focused on testing the `Abort` terminator.

One side-note... The Span for the "hidden else" starts as a 0-length span right after the right brace, if I'm not mistaken. But code region counts versus line counts are confusing for 0-length regions, so I always expend them by 1 character, preferably to the right (in the typical case, after the brace) UNLESS it's at the end of the line, in which case I have to expand it to the left (which is the right brace). 

You can't see it because there are no more code regions on that line, and the code block ends at the semicolon after the `might_abort()` call.

If there was no semicolon (say, if might_abort() returned a value), then there would be multiple code regions on that line, and you would see the one character highlighted. I'm not sure that would help clarify things anymore than in it's existing state though.

tmandry replied:
Yeah, I was able to infer what it was doing, I just found it confusing at first. Your explanation was helpful to understand _why_ doing this is a good idea, thanks.

richkadel replied:
I went ahead and expanded this example, and added comments and notes to help explain the counter for the implicit `else`.

richkadel replied:
Great news here! I changed how the Span is set for implicit `else` when lowering from AST, and it works! No more special case for `Goto`. Lots of related issues resolved.

richkadel commented:
> I think the most important tweak would be to focus on how braces behave, and try to make them as consistent as possible across syntactic forms.

@tmandry I think there may be at least one or two improvements I can make, but while I do agree with you that some of the results are hard to interpret, there are cases that are both hard to interpret and still correct. (Maybe I can't make them any easier to interpret without making the coverage results inaccurate or incomplete.)

In those cases, I don't think we should sacrifice coverage accuracy/completeness just to make the results "appear" more intuitive (and I think you'd agree), but I also don't know if there's much I can do to make them more clear.

Maybe add comments at points in the examples where we see confusing results?

richkadel commented on src/test/run-make-fulldeps/coverage-reports-base/expected_show_coverage.panic_unwind.txt:
```diff
@@ -0,0 +1,50 @@
+   17|      0|            might_panic(true);
+   18|     10|        } else if countdown < 5 {
+   19|      3|            might_panic(false);
+   20|     15|        }
```
I just noticed this incorrect line count, `15`, here and at the same point in the `panic_unwind.rs` test.

This is wrong of course.

I suspect I have overlapping coverage regions with different counters, and they are adding up. (Best guess for now.) I'll look into it.

I think it may be related to how I implement the "hidden else", discussed in my previous reply to @tmandry in `abort.rs`. I may need to rethink what to do if the span is at the end of a line, and possible expand to the first character in the next line, if not at the end of the body.

(FYI, the expansion to the left is actually what I want for the end of the body. This is how I count function `Return`s. So I don't want to break that.)

richkadel replied:
I fixed this when I fixed the `Goto` implicit else span, and remove the special case handling of `Goto` (which didn't work that well in some cases... like this one).

richkadel commented on src/test/run-make-fulldeps/coverage-reports-base/expected_show_coverage.async.txt:
```diff
@@ -0,0 +1,66 @@
+   11|       |        y if f().await == y => (),
+   12|       |        _ => (),
+   13|       |    }
+   14|      1|}
```
I think I'd like to count the opening brace or signature, as you suggest. I'll look into that.

There's a second issue here:

Why do the generator bodies show no counts? (Not even 0.)

I need to understand that and see if it can be fixed.

If I counted the bodies, then we SHOULD see something like:

```
    5|      1|async fn f() -> u8 { 1 }
                                                  ^0
```

As it is, though, there's no way to tell that this async function was actually NOT executed.

tmandry replied:
Interesting. I wonder if the counters are getting optimized out somehow.

richkadel replied:
Partial correction. The body of `f()` actually was executed (because the test blocks on `i()`, and `i()` `await`s `f()`.

But `g()` is a similar example of the point I was trying to make.

From the debug logs, I can see the generator/closure MIR for the body of `g()` is never passed to Rust's codegen phase. 

(FYI, -Clink-dead-code does not make a difference either; I think that flag is only used by the target linker.)

I think I may have looked into this (and if so, it was a hard problem) but it would be nice to find any MIR's skipped at codegen, and still inject something in the coverage map for the function body to, hopefully, produce line counts of 0 for the body.

richkadel replied:
@tmandry @wesleywiser - I'm a little unsure how to proceed here.

I'll try my best to summarize my understanding and what I think I want to do but don't know how:

1. Given a `Crate`, `rustc_mir::monomorphize::collector` finds the "root" `DefId`s (via the `RootCollector`, a HIR `ItemLikeVisitor`), and traverses them to generate a set of `MonoItem`s. This is where the async closures (`Generator`s) are getting dropped if they are not going to be called. (From debug logs, `i::{closure#0}()` and `f::{closure#0}()` are both logged, because they are going to be called, but the other async functions' closures are not mentioned at all. There is no "path" from the roots (only `main()` in this case) to those closures.

2. It seems to me I can make another `ItemLikeVisitor`, find all of the `hir::ItemKindFn`s, get their `DefId`s, and look them up in the set of `MonoItem`s. If a `DefId` is not found among the `MonoItem`s, then it will not be codegen'ed, and therefore won't have coverage.

3. At that point, I want to add at least one `Zero` counter with span (code region) for the span of the closure/generator, for each missing `DefId`, to the `rustc_codegen_ssa::coverageinfo::map::FunctionCoverage` (the collection of counters and coderegions built up during the codegen phase, and eventually written out as the "Coverage Map". I would do this with `FunctionCoverage::add_counter(self, counter, region)`.

Here's where I get confused...

The `FunctionCoverage` is stored in a hash map in `CrateCoverageContext`, and that object is constructed as a field in `CodegenCx`, but `CodegenCx` appears to be constructed for a *specific `CodegenUnit`*

IIUC, codegen of a given crate can be partitioned across multiple `CodegenUnit`s, and each `CodegenUnit` will have only a subset of the crate's `MonoItem`s; so if that's true:

1. Given a `DefId` in some `Crate`, even if the `DefId` is not be found among the `MonoItem`s of one `CodegenUnit`, the `DefId` might still be found among the `MonoItem`s of one of the other`CodegenUnit`s for that `Crate`. If so, that prevents me being able to check for missing `DefIds` when I'm starting to codegen the Coverage Map, because the Coverage Map finalization happens within the `CodegenCx` of only one `CodegenUnit`.

2. But if I search for a `DefId` among all `MonoItem`s of all `CodegenUnit`s, and don't find it, how do I know which `CodegenCx` to use, to add the missing Counter and code region for that function?

What's even more baffling to me is, it looks to me like a CoverageMap is codegenned for each `CodegenCx`, and assuming there are more than one, then the code appears to be set up to generate multiple coverage maps. I don't really get how that would work (or why it appears to work now), but I assume it does work since evidence suggests as much.

So I guess I just don't understand something, and hopefully the solution (to adding counters and code regions for functions/closures that are missing from the MonoItems is easier than it appears.

richkadel replied:
The discussion continued at https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/counters.20for.20uncalled.20functions

richkadel replied:
I couldn't simply inject code regions for functions not codegenned, as discussed in the PR for my attempted changes in draft PR #79392, but I was able to make other changes to improve this.

I'm now adding a coverage code region for the function signature, up to (but not including) the opening brace. The existing code for combining spans helps simplify (I think in a good way) how the function signature span can continue into the function body, if no other spans interfere. I think the results look good.

I also addressed showing `0` coverage for closures and `async` function bodies by injecting what the Coverage Mapping Format calls a "gap region". There are additional notes on this, and some non-intuitive results (but not wrong) occasionally. Given the trade-offs, this is better for ensuring missing coverage is identified.

And, as I showed in my earlier comment tonight, `-C link-dead-code` *does* make a difference for some unused functions. (Just pointing that out, since I discussed it here as well.)

richkadel replied:
I made a different change, backing out my original approach to closures, and I'm now adding coverage for any unused MIR. This addresses missing coverage for async functions, closures, and unused regular functions.

tmandry commented:
@bors delegate+

bors commented:
:v: @richkadel can now approve this pull request
<!-- homu: {"type":"Delegated","delegator":"tmandry","delegate":"richkadel"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #79441) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

richkadel commented:
Some significant coverage count improvements while also simplifying the code a bit, removing some special case handling.

richkadel commented:
I learned some things while reviewing these changes, and managed to simplify things even further, removing the benefits of enforcing `-Clink-dead-code`. That means I have no need to test differently on MSVC, so I was able to remove the duplicate tests (testing `link-dead-code` on and off). So I removed half of the `run-make-fulldeps/coverage-*` tests.

I did run into a couple of weird anomalies, which I had to account for in the `coverage-reports` Makefile (just to make sure the output stayed consistent, for now). Adding `--emit llvm-ir`, for some reason, changes the output  now (or rather, if I don't add that option, the output is a bit different compared to how it was before this commit). It only affects a couple of tests (`async.rs` and `closures.rs`, I believe, both of which have closures).

The other is, a couple of tests create coverage results from built-in Rust macros, that now show up in `llvm-cov show` results. I added a filename filter to exclude them. The results don't appear helpful, but they aren't wrong, and it's not unusual to filter some boilerplate from coverage results.

I'm assuming that the new procedure in `mapgen.rs` for adding missing coverage causes both of these issues, but I don't know why at this time.

richkadel commented:
@tmandry - Although you gave me permission to approve this already, I do think the changes here are substantial enough to warrant another look. 

Good news is, I removed more that I added (in the normal code base) AND removed half of the tests (due to removing the need for `-Clink-dead-code`. 

Can you take another look?

richkadel commented:
> The other is, a couple of tests create coverage results from built-in Rust macros, that now show up in `llvm-cov show` results. I added a filename filter to exclude them. The results don't appear helpful, but they aren't wrong, and it's not unusual to filter some boilerplate from coverage results.

I'm working on what I think will be a good fix for this issue. I don't know if it will also solve the other issue yet.

richkadel commented:
> I'm working on what I think will be a good fix for this issue. I don't know if it will also solve the other issue yet.

Fixed the issue with unnecessarily showing uncalled functions in files not already part of the coverage file set, allowing me to remove one of the two new workarounds in the Makefile.

This did not fix the other issue related to differences when not including `--emit llvm`. I'm going to take a closer look at that issue.

richkadel commented:
I'm narrowing in on the fixes for the remaining issue. I can't explain why/how including `--emit llvm-ir` works around the issue, but so far I believe there is at least one bug I will be fixing (addressing part of the anomaly), and I hope to find a way to fix the rest of it very soon.

richkadel commented:
@tmandry - I was able to work out how to find and add unreachable code regions without assuming anything about the CGU partitioning. The results are as good as I had hoped (and improved even over the last PR because I'm now adding all code regions from the MIR analysis instead of just one big one for the entire body. This will have advantages at analysis time, plus it looks more consistent when comparing unused functions/closures to used functions.

I pushed one more commit with these changes, and I made a copy of the branch in case you want to see what the last change was:

https://github.com/richkadel/rust/tree/llvm-coverage-counters-2.0.5-b4-squash

But I'm going to start squashing now, so *this* branch (PR) will change shortly.

richkadel commented:
@tmandry - OK, done with squashing, and I inserted a new commit in between the last one you reviewed and the final commit, changing the "-base" path tests to remove "-base". So comparing the last commit (commit 3) to the inserted second commit, search for:

src/test/run-make-fulldeps/coverage-reports/expected_show_coverage.abort.txt

as a starting point.

You can see the improvements. 

In a few cases I added some extra test cases and some comments that throw off the line numbering, sorry about that.

tmandry commented on src/test/run-make-fulldeps/coverage-reports/expected_show_coverage.abort.txt:
```diff
@@ -2,34 +2,69 @@
-   34|       |//      (on Linux at least).
+   20|       |        // See discussion (below the `Notes` section) on coverage results for the closing brace.
+   21|     10|        if countdown < 5 { might_abort(false); } // Counts for different regions on one line.
+                                       ^4                     ^6
```
Pointing one past the closing brace looked a bit odd to me at first, but I'm starting to like it more now that I think about it as having "skipped past the if block". (Like you said, there's not a 100% intuitive way to expose all the information here, so we just have to pick something.)

tmandry commented on src/doc/unstable-book/src/compiler-flags/source-based-code-coverage.md:
```diff
@@ -1,14 +1,14 @@
 # `source-based-code-coverage`
 
-The tracking issue for this feature is: [#79121](https://github.com/rust-lang/rust/issues/79121).
+The tracking issue for this feature is: [#79121].
```
These doc changes all look really helpful, thanks!

tmandry commented on src/test/run-make-fulldeps/coverage-reports/expected_show_coverage.async.txt:
```diff
@@ -2,65 +2,130 @@
+   10|       |    }
+   11|      1|}
+   12|       |
+   13|      0|async fn d() -> u8 { 1 }
```
I'm looking for a one-line function that gets called but whose future is never polled.. I'd expect to see a line count of 1 with a 0 counter inside the body.

richkadel replied:
I've added `m()`:

```
   90|       |
   91|      1|async fn m(x: u8) -> u8 { x - 1 }
                                      ^0
   92|       |
   93|      1|fn main() {
   94|      1|    let _ = g(10);
   95|      1|    let _ = h(9);
   96|      1|    let mut future = Box::pin(i(8));
   97|      1|    j(7);
   98|      1|    l(6);
   99|      1|    let _ = m(5);
  100|      1|    executor::block_on(future.as_mut());
  101|      1|}
```

richkadel replied:
Done

tmandry commented on src/test/run-make-fulldeps/coverage-reports/expected_show_coverage.conditions.txt:
```diff
@@ -38,69 +38,53 @@
+   45|      1|        if countdown > 7 {
+   46|      1|            countdown -= 4;
+   47|      1|        }
+   48|      0|        else if countdown > 2 {
```
With the count on the closing brace above being 1, I'd expect to see 1 here too since the condition must be evaluated.

richkadel replied:
I believe this is correct, but it may be easier to see in `--color`:

<img width="471" alt="Screen Shot 2020-12-01 at 9 59 04 PM" src="https://user-images.githubusercontent.com/3827298/100834357-9ba7df00-3420-11eb-9042-1535abdf293f.png">

The condition on line 45 is `true`, and the block surrounding line 46 is executed; so there's no need to evaluate the condition on line 48.

tmandry replied:
I see, so do we always count the closing brace itself as part of the if body (whether or not it has an `else`?) I was thinking we didn't include the brace, and that the count on that line was for the "condition was false" branch.

richkadel replied:
Yes... and the "condition was false" is indicated by the coverage of the block and statements _after_ the `else` (if any). I **think** this is the most accurate it can be.

This was part of the improvement to get rid of `Goto` special case handling, and as discussed above, the counter carat `^0` for an implicit else appears _after_ the closing brace now _because_ the closing brace is part of the `Block` of the previous `if`.

I also tried to show this by the unorthodox formatting, showing various pieces of the syntax on different lines, occasionally, to document through testing how the coverage covers each part (since the test result don't have the benefit of colors).

tmandry replied:
Got it, thanks.

tmandry commented on src/test/run-make-fulldeps/coverage-reports/expected_show_coverage.conditions.txt:
```diff
@@ -38,69 +38,53 @@
+   50|      0|                countdown = 0;
+   51|      0|            }
+   52|      0|            countdown -= 5;
+   53|       |        } else {
```
Also expecting to see a count here

richkadel replied:
I'm not really sure I follow, but I think the coverage spans are accurate to what is being executed.

Maybe the example in color helps clarify what you're concerned about here? That example also includes the lines you reference in this comment.

tmandry commented on src/test/run-make-fulldeps/coverage-reports/expected_show_coverage.conditions.txt:
```diff
@@ -38,69 +38,53 @@
+   68|      0|            countdown = 0;
+   69|      0|        }
+   70|      0|        countdown -= 5;
+   71|       |    } else {
```
ditto (and below)

richkadel replied:
To me, these all look completely accurate. Here's the color version including these lines...

<img width="443" alt="Screen Shot 2020-12-01 at 10 16 19 PM" src="https://user-images.githubusercontent.com/3827298/100835739-ef1b2c80-3422-11eb-8e01-a2eb6f6cf47f.png">

* `countdown > 7` is checked (`false`)
* `countdown > 2` is checked (`false`)
* the `else` block is executed

tmandry replied:
I see, I guess because there is an `else` block, all the information is still there. What confused me is the inconsistency with `if` expressions that have no `else` block. If we always included the span after the closing brace as part of the `else` block it would be more consistent and self-explanatory, in my opinion.

richkadel replied:
I get your point, but that's not how the HIR spans set it up. (But that could be changed, as I did here for implicit else.) I'm not sure if your suggestion is best in the long run... or if there are other implications (e.g., with if-else-if-else.). Maybe.

tmandry replied:
Ah so I think doing something like what you did for all if exprs (not just those that have no `else`) would give more consistent looking output, personally, and I think it can work fine with chained `else if`s.

richkadel replied:
Yeah perhaps. I'm still concerned about unexpected side effects.

tmandry commented on src/test/run-make-fulldeps/coverage-reports/expected_show_coverage.partial_eq.txt:
```diff
@@ -2,15 +2,15 @@
     3|       |
     4|      2|#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
-                                                    ^1       ^1
+                       ^0            ^0      ^0 ^0  ^1       ^0 ^0^0
```
Cool that we get coverage for these, but what's going on with the second counters under PartialEq and Ord?

richkadel replied:
I don't really know how they get expanded.

I'll add a FIXME comment and issue. I don't think this is high priority, but it might be nice to clean up in the future.

richkadel replied:
Issue #79626

richkadel replied:
Added issue and referenced it in a comment in the test.

tmandry reviewed (approved):
This seems like a massive improvement overall!

I found some counts that don't seem quite right. If they don't seem fixable in a timely manner you can always file an issue for us to follow up on later. I'm also willing to merge this without any major changes and do follow up changes in new PRs.

richkadel commented on compiler/rustc_codegen_llvm/src/coverageinfo/mapgen.rs:
```diff
@@ -228,3 +235,137 @@ fn save_function_record(
+///    function (must not have type parameters) because the coverage tools will get confused
+///    if the codegenned function has more than one instantiation and additional `CodeRegion`s
+///    attached to only one of those instantiations.
+fn add_unreachable_coverage<'tcx>(
```
adding a comment referencing new issue #79622

richkadel replied:
Done

richkadel commented on compiler/rustc_middle/src/query/mod.rs:
```diff
@@ -346,6 +346,21 @@ rustc_queries! {
+            cache_on_disk_if { key.is_local() }
+        }
+
+        /// Returns the `CodeRegions` for a function that has instrumented coverage, in case the
+        /// function was optimized out before codegen, and before being added to the Coverage Map.
```
I'm adding some comments explaining my justification for the current approach. I think the penalty is low, and the downsides of a single code region are potentially significant.

richkadel replied:
Done

richkadel commented on compiler/rustc_codegen_llvm/src/coverageinfo/mapgen.rs:
```diff
@@ -228,3 +235,137 @@ fn save_function_record(
+///
+/// *HOWEVER* the codegenned `DefId`s are partitioned across multiple `CodegenUnit`s (CGUs), and
+/// this function is processing a `function_coverage_map` for the functions (`Instance`/`DefId`)
+/// allocated to only one of those CGUs. We must NOT inject any "Unreachable" functions's
+/// `CodeRegion`s more than once, so we have to pick which CGU's `function_coverage_map` to add
+/// each "Unreachable" function to.
```
I understand your question, and in fact it's one of the examples that I just put in comment about why I'm adding each individual code region for `Unreachable`s, instead of one broad one.

If a function is covered when compiling one crate, but considered `Unreachable` from another, the overlapping but exactly equal code spans do not conflict. The `Unreachable` coverages just add `0` (no visible affect in coverage analysis).

I'll try to construct a test.

richkadel replied:
Done. See new `uses_crate.rs` (with `deps/used_crate.rs`) test and results.

richkadel commented on compiler/rustc_mir/src/transform/coverage/mod.rs:
```diff
@@ -521,10 +495,13 @@ fn make_code_region(
 }
 
-fn hir_body<'tcx>(tcx: TyCtxt<'tcx>, def_id: DefId) -> &'tcx rustc_hir::Body<'tcx> {
+fn fn_sig_and_body<'tcx>(
```
Added #79625

richkadel replied:
Done and referenced in a comment.

richkadel commented on compiler/rustc_mir/src/transform/coverage/spans.rs:
```diff
@@ -345,28 +352,28 @@ impl<'a, 'tcx> CoverageSpans<'a, 'tcx> {
+        // closure's `Return` is the only one that will be counted. This provides adequate
+        // coverage, and more intuitive counts. (Avoids double-counting the closing brace of the
+        // function body.)
+        let body_ends_with_closure = if let Some(last_covspan) = refined_spans.last() {
```
I don't think that's syntactically possible, but if it was, then I think the solution here still applies.

With a closure, I believe there will also be some tokens separating the end of the closure and the end of the function.

That's not the case for the async functions.

tmandry replied:
What I meant was, what happens when we instrument the closure itself? It seems like the body of the closure does indeed end with a closure :)

richkadel replied:
Oh! The MIR for the closure itself doesn't consider itself a closure. The check only applies in the MIR/function that declares/defines the closure.

richkadel commented:
@tmandry - I think I've addressed all of your feedback with a change, if possible, or an answer. Thanks for the detailed feedback again!

I'm going to reconfirm the tests, bless changes or added tests, and upload the changes. I assume you're still OK with me submitting this to bors on your behalf, once these changes are uploaded correctly?

richkadel commented on compiler/rustc_mir/src/transform/coverage/graph.rs:
```diff
@@ -147,13 +137,22 @@ impl CoverageGraph {
+
+                // The following `TerminatorKind`s are either not expected outside an unwind branch,
+                // or they should not (under normal circumstances) branch. Coverage graphs are
+                // simplified by assuring coverage results are accurate for well-behaved programs.
```
Done

richkadel commented on compiler/rustc_mir/src/transform/coverage/spans.rs:
```diff
@@ -345,28 +352,28 @@ impl<'a, 'tcx> CoverageSpans<'a, 'tcx> {
-            }
+
+        // Async functions wrap a closure that implements the body to be executed. The enclosing
+        // function is initially called, posts the closure to the executor, and returns. To avoid
```
Done

richkadel commented:
@bors r=tmandry

bors commented:
:pushpin: Commit de62f04869fccd608efafd87d8bac52569611e80 has been approved by `tmandry`

<!-- @bors r=tmandry de62f04869fccd608efafd87d8bac52569611e80 -->
<!-- homu: {"type":"Approved","sha":"de62f04869fccd608efafd87d8bac52569611e80","approver":"tmandry"} -->

richkadel commented:
@bors r=tmandry

bors commented:
:pushpin: Commit c47c88c71510793c82e874901b114752b59252e5 has been approved by `tmandry`

<!-- @bors r=tmandry c47c88c71510793c82e874901b114752b59252e5 -->
<!-- homu: {"type":"Approved","sha":"c47c88c71510793c82e874901b114752b59252e5","approver":"tmandry"} -->

bors commented:
:hourglass: Testing commit c47c88c71510793c82e874901b114752b59252e5 with merge 1e875c785678a721e1d118e274f75fddf5233fde...
<!-- homu: {"type":"BuildStarted","head_sha":"c47c88c71510793c82e874901b114752b59252e5","merge_sha":"1e875c785678a721e1d118e274f75fddf5233fde"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1485782936)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1485782936","builder_name":"checks-actions"} -->

richkadel commented:
@tmandry - I added the new test with multiple crates (multiple files), and ran into another Windows-specific problem.

On Windows, for some reason, the `llvm-cov show` results show the two files in different orders!

It looks like I also have path name issues (`\` vs `/`).

But I implemented a workaround that I think is "reasonable" in the near term, and meets the testing expectations. We can look at a longer term solution later, but given the narrow scope of the issue at this time, it's not worth the effort to get so elaborate right now.

```
diff -u --strip-trailing-cr expected_show_coverage.uses_crate.txt "/d/a/rust/rust/build/x86_64-pc-windows-msvc/test/run-make-fulldeps/coverage-reports/coverage-reports"/actual_show_coverage.uses_crate.txt || \
	( grep -q '^\/\/ ignore-llvm-cov-show-diffs' ../coverage/uses_crate.rs && \
		>&2 echo 'diff failed, but suppressed with `// ignore-llvm-cov-show-diffs` in ../coverage/uses_crate.rs' \
	) || \
	( >&2 echo 'diff failed, and not suppressed without `// ignore-llvm-cov-show-diffs` in ../coverage/uses_crate.rs'; \
		false \
	)
--- expected_show_coverage.uses_crate.txt	2020-12-02 09:56:05.285910400 +0000
+++ /d/a/rust/rust/build/x86_64-pc-windows-msvc/test/run-make-fulldeps/coverage-reports/coverage-reports/actual_show_coverage.uses_crate.txt	2020-12-02 10:56:42.639202900 +0000
@@ -1,4 +1,16 @@
-../coverage/used_crate/mod.rs:
+..\coverage\uses_crate.rs:
+    1|       |#![allow(unused_assignments, unused_variables)]
+    2|       |
+    3|       |mod used_crate;
+    4|       |
+    5|      1|fn main() {
+    6|      1|    used_crate::used_function();
+    7|      1|    let some_vec = vec![1, 2, 3, 4];
+    8|      1|    used_crate::used_generic_function(&some_vec);
+    9|      1|    used_crate::used_twice_generic_function(some_vec);
+   10|      1|}
+
+..\coverage\used_crate\mod.rs:
     1|       |#![allow(unused_assignments, unused_variables)]
     2|       |
     3|       |use std::fmt::Debug;
@@ -55,15 +67,3 @@
    42|      0|    }
    43|      0|}
 
-../coverage/uses_crate.rs:
-    1|       |#![allow(unused_assignments, unused_variables)]
-    2|       |
-    3|       |mod used_crate;
-    4|       |
-    5|      1|fn main() {
-    6|      1|    used_crate::used_function();
-    7|      1|    let some_vec = vec![1, 2, 3, 4];
-    8|      1|    used_crate::used_generic_function(&some_vec);
-    9|      1|    used_crate::used_twice_generic_function(some_vec);
-   10|      1|}
-
make[1]: Leaving directory '/d/a/rust/rust/src/test/run-make-fulldeps/coverage-reports'
```

richkadel commented:
The changes are limited to tests that only run in CI. It should be OK to push to bors now:

@bors r=tmandry

bors commented:
:pushpin: Commit 00117d712189ca703b481b471e11bcc742f0346e has been approved by `tmandry`

<!-- @bors r=tmandry 00117d712189ca703b481b471e11bcc742f0346e -->
<!-- homu: {"type":"Approved","sha":"00117d712189ca703b481b471e11bcc742f0346e","approver":"tmandry"} -->

tmandry commented:
@bors rollup=never given the size

tmandry commented on src/test/run-make-fulldeps/coverage/uses_crate.rs:
```diff
@@ -0,0 +1,10 @@
+#![allow(unused_assignments, unused_variables)]
+
+mod used_crate;
```
This is a different module in the same crate. To test two crates, you need two rustc invocations.

richkadel replied:
Done

tmandry commented:
@bors r- since the crate test isn't quite doing what's advertised

tmandry commented:
@bors r+

bors commented:
:pushpin: Commit ad4631aa712afb1259a4bfd0ce9916ee9d255508 has been approved by `tmandry`

<!-- @bors r=tmandry ad4631aa712afb1259a4bfd0ce9916ee9d255508 -->
<!-- homu: {"type":"Approved","sha":"ad4631aa712afb1259a4bfd0ce9916ee9d255508","approver":"tmandry"} -->

bors commented:
:hourglass: Testing commit ad4631aa712afb1259a4bfd0ce9916ee9d255508 with merge 7fd5d63a550eaadbb7540dd8533c346169f2eeef...
<!-- homu: {"type":"BuildStarted","head_sha":"ad4631aa712afb1259a4bfd0ce9916ee9d255508","merge_sha":"7fd5d63a550eaadbb7540dd8533c346169f2eeef"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1490669528)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1490669528","builder_name":"checks-actions"} -->

richkadel commented:
As we've seen before, `llvm-cov show` sometimes displays multiple instantiations of generic types in a non-deterministic order, particularly on Windows.

I have a test directive to ignore the `llvm-cov show` results if necessary, but I haven't actually needed to use it lately, and it's very coarse grained (ignoring the entire file, testing only the `.json` report instead).

Good news is, I think I can set it up to ONLY ignore the injected labels/headers introducing the instantiations, using `diff --ignore-matching-lines=RE`. I tested it and it works. I checked Linux, Mac, and Windows, and that `diff` option is available on all three platforms.

I can probably get rid of the `.json` files, and the test directive that's no longer used (in a future PR). "TODO(richkadel)" comments have been added.

FYI, while reviewing the error, I also noticed I missed a line when appending `/lib` to `$(SOURCEDIR)` in both Makefiles. It broke the `grep` but still continued the test. I fixed that.

I also noticed that I forgot to refresh the blessed file when fixing the `test tidy` error (on the same test that broke anyway, with the generic instantiation issue). So that's also fixed now.

```
--- expected_show_coverage.uses_crate.txt	2020-12-03 04:15:23.590553900 +0000
+++ /d/a/rust/rust/build/x86_64-pc-windows-msvc/test/run-make-fulldeps/coverage-reports/coverage-reports/actual_show_coverage.uses_crate.txt	2020-12-03 05:16:24.496488800 +0000
@@ -19,12 +19,12 @@
    18|      2|    println!("used_only_from_bin_crate_generic_function with {:?}", arg);
    19|      2|}
   ------------------
-  | used_crate::used_only_from_bin_crate_generic_function::<&str>:
+  | used_crate::used_only_from_bin_crate_generic_function::<&alloc::vec::Vec<i32>>:
   |   17|      1|pub fn used_only_from_bin_crate_generic_function<T: Debug>(arg: T) {
   |   18|      1|    println!("used_only_from_bin_crate_generic_function with {:?}", arg);
   |   19|      1|}
   ------------------
-  | used_crate::used_only_from_bin_crate_generic_function::<&alloc::vec::Vec<i32>>:
+  | used_crate::used_only_from_bin_crate_generic_function::<&str>:
   |   17|      1|pub fn used_only_from_bin_crate_generic_function<T: Debug>(arg: T) {
   |   18|      1|    println!("used_only_from_bin_crate_generic_function with {:?}", arg);
   |   19|      1|}
@@ -103,8 +103,8 @@
    66|       |//  | Unexecuted instantiation: <some function name here>
    67|       |//  ------------------
    68|       |//
-   69|       |// Note, the function name shown in the error seems to change depending on the structure of the code,
-   70|       |// for some reason, including:
+   69|       |// Note, the function name shown in the error seems to change depending on the structure of the
+   70|       |// code, for some reason, including:
    71|       |//
    72|       |// * used_crate::used_from_bin_crate_and_lib_crate_generic_function::<&str>
    73|       |// * used_crate::use_this_lib_crate
make[1]: Leaving directory '/d/a/rust/rust/src/test/run-make-fulldeps/coverage-reports'

------------------------------------------
stderr:
------------------------------------------
grep: ../coverage/used_crate.rs: No such file or directory
```

richkadel commented:
@bors r=tmandry

bors commented:
:pushpin: Commit 2c99600a8679615f601aa010825012e9b46fe59b has been approved by `tmandry`

<!-- @bors r=tmandry 2c99600a8679615f601aa010825012e9b46fe59b -->
<!-- homu: {"type":"Approved","sha":"2c99600a8679615f601aa010825012e9b46fe59b","approver":"tmandry"} -->

bors commented:
:hourglass: Testing commit 2c99600a8679615f601aa010825012e9b46fe59b with merge 821363e61eff90b9a4f8fbceb912ffc374b9790d...
<!-- homu: {"type":"BuildStarted","head_sha":"2c99600a8679615f601aa010825012e9b46fe59b","merge_sha":"821363e61eff90b9a4f8fbceb912ffc374b9790d"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1491714247)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1491714247","builder_name":"checks-actions"} -->

richkadel commented:
```
Some tests failed in compiletest suite=run-make-fulldeps mode=run-make host=x86_64-apple-darwin target=x86_64-apple-darwin

failures:

---- [run-make] run-make-fulldeps/coverage-reports stdout ----

error: make failed
status: exit code: 2
command: "make"
stdout:
------------------------------------------
# Compile the test library with coverage instrumentation
DYLD_LIBRARY_PATH="/Users/runner/work/rust/rust/build/x86_64-apple-darwin/test/run-make-fulldeps/coverage-reports/coverage-reports:/Users/runner/work/rust/rust/build/x86_64-apple-darwin/stage2/lib:" '/Users/runner/work/rust/rust/build/x86_64-apple-darwin/stage2/bin/rustc' --out-dir /Users/runner/work/rust/rust/build/x86_64-apple-darwin/test/run-make-fulldeps/coverage-reports/coverage-reports -L /Users/runner/work/rust/rust/build/x86_64-apple-darwin/test/run-make-fulldeps/coverage-reports/coverage-reports  ../coverage/lib/used_crate.rs \
			$( grep -q '^\/\/ require-rust-edition-2018' ../coverage/lib/used_crate.rs && echo "--edition=2018" ) \
			--crate-type rlib \
			-Zinstrament-coverage

------------------------------------------
stderr:
------------------------------------------
error: unknown debugging option: `instrament-coverage`

make: *** [used_crate] Error 1

------------------------------------------



failures:
    [run-make] run-make-fulldeps/coverage-reports

test result: FAILED. 198 passed; 1 failed; 16 ignored; 0 measured; 0 filtered out; finished in 122.30s
```

richkadel commented:
So this is only on:

auto (x86_64-apple, ./x.py --stage 2 test, --build=x86_64-apple-darwin --enable-sanitizers --enab...

All tests appear to have passed on all other platforms.

I don't see how the command is bad, especially only for Mac.

If it isn't, then somehow RUSTC no longer understands `-Zinstrument-coverage` at this point?

The command did change slightly:

```diff
	$(RUSTC) $(SOURCEDIR)/$@.rs \
-			-Zinstrument-coverage \
-			-Clink-dead-code=$(LINK_DEAD_CODE)
+			$$( grep -q '^\/\/ require-rust-edition-2018' $(SOURCEDIR)/$@.rs && \
+				echo "--edition=2018" \
+			) \
+			-L "$(TMPDIR)" \
+			-Zinstrument-coverage
```

And the command is now run two different ways, the first one for library crates (that's the first one encountered, and is the failure above:

```
	$(RUSTC) $(SOURCEDIR)/lib/$@.rs \
			$$( grep -q '^\/\/ require-rust-edition-2018' $(SOURCEDIR)/lib/$@.rs && \
				echo "--edition=2018" \
			) \
			--crate-type rlib \
			-Zinstrument-coverage
```


```

richkadel commented:
> error: unknown debugging option: `instrament-coverage`

Wait! That's a typo!

(But how did it only fail on apple?)

richkadel commented:
I don't see that word, `instrament`, in the code base or in my commit at all.

What kind of bug is this?

richkadel commented:
? Spurious ?

richkadel commented:
@bors retry

bors commented:
:hourglass: Testing commit 2c99600a8679615f601aa010825012e9b46fe59b with merge cfd64d0875173833732b8c0b2c9ca92bc0d0b1bb...
<!-- homu: {"type":"BuildStarted","head_sha":"2c99600a8679615f601aa010825012e9b46fe59b","merge_sha":"cfd64d0875173833732b8c0b2c9ca92bc0d0b1bb"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1494635575)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1494635575","builder_name":"checks-actions"} -->

richkadel commented:
@bors retry

richkadel commented:
See https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/x86_64-apple-darwin.20CI.20randomly.20injecting.20typos.20in.20Makefile.3F

richkadel commented:
@bors r=tmandry

bors commented:
:pushpin: Commit dc4bd9067eda1df118fc48cb6b828d72e916e1bb has been approved by `tmandry`

<!-- @bors r=tmandry dc4bd9067eda1df118fc48cb6b828d72e916e1bb -->
<!-- homu: {"type":"Approved","sha":"dc4bd9067eda1df118fc48cb6b828d72e916e1bb","approver":"tmandry"} -->

yaahc commented on src/doc/unstable-book/src/compiler-flags/source-based-code-coverage.md:
```diff
@@ -156,14 +168,77 @@ There are four statistics tracked in a coverage summary:
+$ cargo cov -- report \
+    --use-color --ignore-filename-regex='/.cargo/registry' \
+    --instr-profile=json5format.profdata \
+    target/debug/deps/lib-30768f9c53506dc5 \
+    target/debug/deps/json5format-fececd4653271682
```
my experience in https://github.com/ZcashFoundation/zebra/pull/1293 indicates that this will not produce the proper coverage report. I didn't quite get to the bottom of exactly what was happening but when I tried specifying the bins as a list of space separated arguments it would lose some of the coverage reports, and ordering mattered. My guess is it was ignoring everything after the first, or overwriting them somehow. In order to fix it I had to prefix _every_ bin with `-object`.

https://github.com/ZcashFoundation/zebra/pull/1293/commits/f4b1cd2bd0ec8958fc1e229fe5431381a1fc5852 This is the PR where I did it, the fix being adding the `printf` when we output the file names.

It might also be a good idea to suggest how to extract the list of binaries that were ran by `cargo-test`, or link to examples that do so.

```bash
cargo test --no-run --message-format=json \
    | jq -r "select(.profile.test == true) | .filenames[]" \
    | grep -v dSYM -
```

richkadel replied:
Cool! Thanks for the recommendation. I've been struggling to get this PR to land so I may apply these suggestions in a follow-up PR.

richkadel replied:
I've addressed this in #79818.

Thanks again!

bors commented:
:hourglass: Testing commit dc4bd9067eda1df118fc48cb6b828d72e916e1bb with merge 6513f5029183471993bab4b05ae5f5433f461b08...
<!-- homu: {"type":"BuildStarted","head_sha":"dc4bd9067eda1df118fc48cb6b828d72e916e1bb","merge_sha":"6513f5029183471993bab4b05ae5f5433f461b08"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1496918682)
Approved by: tmandry
Pushing 6513f5029183471993bab4b05ae5f5433f461b08 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"tmandry","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1496918682"},"merge_sha":"6513f5029183471993bab4b05ae5f5433f461b08"} -->
